hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/PulsarSource.java

```diff
@@ -0,0 +1,292 @@
+    if (sourceLimit < Long.MAX_VALUE) {
+      return sourceLimit;
+    } else {
+      return props.getLong(Config.PULSAR_SOURCE_MAX_RECORDS_PER_BATCH_THRESHOLD_PROP.key(),
+          Config.PULSAR_SOURCE_MAX_RECORDS_PER_BATCH_THRESHOLD_PROP.defaultValue());
+    }
+  }
+
+  private static String convertToOffsetString(String topic, MessageId startingOffset) {
+    return JsonUtils.topicOffsets(
+        HoodieConversionUtils.mapAsScalaImmutableMap(
+            Collections.singletonMap(topic, startingOffset)));
+  }
+
+  private static void shutdownPulsarClient(PulsarClient client) throws PulsarClientException {
+    client.close();
+    // NOTE: Current version of Pulsar's client (in Pulsar Spark Connector 3.1.1.4) is not
+    //       shutting down event-loop group properly, so we had to shut it down manually
+    try {
+      EventLoopGroup eventLoopGroup = ((PulsarClientImpl) client).eventLoopGroup();
+      if (eventLoopGroup != null) {
+        eventLoopGroup.shutdownGracefully().await();
+      }
+    } catch (InterruptedException e) {
+      // No-op
+    }
+
+    // NOTE: Pulsar clients initialized by the spark-connector, might be left not shutdown
+    //       properly (see above). To work this around we employ "nuclear" option of
+    //       fetching all Pulsar client threads and interrupting them forcibly (to make them
+    //       shutdown)
+    collectActiveThreads().stream().sequential()
+        .filter(t -> t.getName().startsWith("pulsar-client-io"))
+        .forEach(Thread::interrupt);
+  }
+
+  public static class Config {
+    private static final ConfigProperty<String> PULSAR_SOURCE_TOPIC_NAME = ConfigProperty
+        .key("hoodie.deltastreamer.source.pulsar.topic")
+        .noDefaultValue()
+        .withDocumentation("Name of the target Pulsar topic to source data from");
+
+    private static final ConfigProperty<String> PULSAR_SOURCE_SERVICE_ENDPOINT_URL = ConfigProperty
+        .key("hoodie.deltastreamer.source.pulsar.endpoint.service.url")
+        .defaultValue("pulsar://localhost:6650")
+        .withDocumentation("URL of the target Pulsar endpoint (of the form 'pulsar://host:port'");
+
+    private static final ConfigProperty<String> PULSAR_SOURCE_ADMIN_ENDPOINT_URL = ConfigProperty
+        .key("hoodie.deltastreamer.source.pulsar.endpoint.admin.url")
+        .defaultValue("http://localhost:8080")
+        .withDocumentation("URL of the target Pulsar endpoint (of the form 'pulsar://host:port'");
+
+    public enum OffsetAutoResetStrategy {
+      LATEST, EARLIEST, FAIL
+    }
+
+    private static final ConfigProperty<OffsetAutoResetStrategy> PULSAR_SOURCE_OFFSET_AUTO_RESET_STRATEGY = ConfigProperty
+        .key("hoodie.deltastreamer.source.pulsar.offset.autoResetStrategy")
+        .defaultValue(OffsetAutoResetStrategy.LATEST)
+        .withDocumentation("Policy determining how offsets shall be automatically reset in case there's "
+            + "no checkpoint information present");
```

Comment:
minor. does not mean, no checkpoint information present. It means that, if offset to consume if not available in source, what strategy to use to reset the consumption offset.

Reply:
Not sure i understand your point: how's it different from the current wording?

Reply:
it wasn't intuitive to me. feel free to take it leave it. I am good either ways